From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PulseMC Team <developer@pulsemc.dev>
Date: Tue, 13 Jan 2026 16:31:08 +0600
Subject: [PATCH] Pulse github patches test


diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index eb06d8f012684845146429832e977e6c1ddcd62b..7e6de4f07dccb9e4284dc704bd7a8a78f5374623 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -75,7 +75,7 @@ import net.minecraft.world.level.storage.LevelData;
 import net.minecraft.world.level.storage.LevelStorageSource;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
-
+// PULSE_MODIFIED
 public class DedicatedServer extends MinecraftServer implements ServerInterface {
     static final Logger LOGGER = LogUtils.getLogger();
     private static final int CONVERSION_RETRY_DELAY_MS = 5000;
@@ -291,15 +291,15 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
             }
             bindAddress = new io.netty.channel.unix.DomainSocketAddress(this.getLocalIp().substring("unix:".length()));
         } else {
-        InetAddress inetAddress = null;
-        if (!this.getLocalIp().isEmpty()) {
-            inetAddress = InetAddress.getByName(this.getLocalIp());
-        }
+            InetAddress inetAddress = null;
+            if (!this.getLocalIp().isEmpty()) {
+                inetAddress = InetAddress.getByName(this.getLocalIp());
+            }
 
-        if (this.getPort() < 0) {
-            this.setPort(properties.serverPort);
-        }
-        bindAddress = new java.net.InetSocketAddress(inetAddress, this.getPort());
+            if (this.getPort() < 0) {
+                this.setPort(properties.serverPort);
+            }
+            bindAddress = new java.net.InetSocketAddress(inetAddress, this.getPort());
         }
         // Paper end - Unix domain socket support
 
@@ -316,6 +316,22 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
             return false;
         }
 
+        // Pulse start - Init Config and Commands
+        dev.pulsemc.api.config.PulseConfig.load();
+        dev.pulsemc.network.PulseMetrics.start();
+        this.server.getCommandMap().register("pulse", "Pulse", new org.bukkit.command.defaults.BukkitCommand("pulse") {
+            private final dev.pulsemc.command.PulseCommand executor = new dev.pulsemc.command.PulseCommand();
+            @Override
+            public boolean execute(org.bukkit.command.CommandSender sender, String commandLabel, String[] args) {
+                return executor.onCommand(sender, this, commandLabel, args);
+            }
+            @Override
+            public java.util.List<String> tabComplete(org.bukkit.command.CommandSender sender, String alias, String[] args) {
+                return executor.onTabComplete(sender, this, alias, args);
+            }
+        });
+        // Pulse end
+
         // CraftBukkit start
         this.server.loadPlugins();
         this.server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.STARTUP);
diff --git a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index b8a4b4cc02a2fc6b70f4b840796eed501aad6239..e6dbed341ed5b414f5d8feb90fa3e03461ec5315 100644
--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -29,7 +29,9 @@ import net.minecraft.util.VisibleForDebug;
 import net.minecraft.util.profiling.Profiler;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
-
+import dev.pulsemc.network.PulseBuffer;
+import io.papermc.paper.util.KeepAlive;
+// PULSE_MODIFIED
 public abstract class ServerCommonPacketListenerImpl implements ServerCommonPacketListener {
     private static final Logger LOGGER = LogUtils.getLogger();
     public static final int LATENCY_CHECK_INTERVAL = 15000;
@@ -39,13 +41,12 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
     protected final MinecraftServer server;
     public final Connection connection; // Paper
     private final boolean transferred;
-    //private long keepAliveTime; // Paper - improve keepalives
-    //private boolean keepAlivePending; // Paper - improve keepalives
-    //private long keepAliveChallenge; // Paper - improve keepalives
+    private long keepAliveTime;
+    private boolean keepAlivePending;
+    private long keepAliveChallenge;
     private long closedListenerTime;
     private boolean closed = false;
-    private volatile int latency; // Paper - improve keepalives - make volatile
-    private final io.papermc.paper.util.KeepAlive keepAlive; // Paper - improve keepalives
+    private int latency;
     private volatile boolean suspendFlushingOnServerThread = false;
     // CraftBukkit start
     public final org.bukkit.craftbukkit.CraftServer cserver;
@@ -59,18 +60,24 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
     public final java.util.Set<String> pluginMessagerChannels;
     // Paper end - retain certain values
 
+    public PulseBuffer pulseBuffer; // Pulse
+    private final KeepAlive keepAlive; // Pulse
+
+
+
     public ServerCommonPacketListenerImpl(MinecraftServer server, Connection connection, CommonListenerCookie cookie) {
         this.server = server;
         this.connection = connection;
-        //this.keepAliveTime = Util.getMillis(); // Paper - improve keepalives
+        this.keepAliveTime = Util.getMillis();
         this.latency = cookie.latency();
         this.transferred = cookie.transferred();
         // Paper start
         this.playerBrand = cookie.brandName();
         this.cserver = server.server;
         this.pluginMessagerChannels = cookie.channels();
-        this.keepAlive = cookie.keepAlive();
         // Paper end
+        this.pulseBuffer = new PulseBuffer(this); // Pulse
+        this.keepAlive = cookie.keepAlive();
     }
 
     // Paper start - configuration phase API
@@ -102,41 +109,13 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
 
     @Override
     public void handleKeepAlive(ServerboundKeepAlivePacket packet) {
-        // Paper start - improve keepalives
-        long now = System.nanoTime();
-        io.papermc.paper.util.KeepAlive.PendingKeepAlive pending = this.keepAlive.pendingKeepAlives.peek();
-        if (pending != null && pending.challengeId() == packet.getId()) {
-            this.keepAlive.pendingKeepAlives.remove(pending);
-
-            io.papermc.paper.util.KeepAlive.KeepAliveResponse response = new io.papermc.paper.util.KeepAlive.KeepAliveResponse(pending.txTimeNS(), now);
-
-            this.keepAlive.pingCalculator1m.update(response);
-            this.keepAlive.pingCalculator5s.update(response);
-
-            this.latency = this.keepAlive.pingCalculator5s.getAvgLatencyMS();
-            return;
-        }
-
-        for (java.util.Iterator<io.papermc.paper.util.KeepAlive.PendingKeepAlive> itr = this.keepAlive.pendingKeepAlives.iterator(); itr.hasNext();) {
-            io.papermc.paper.util.KeepAlive.PendingKeepAlive ka = itr.next();
-            if (ka.challengeId() == packet.getId()) {
-                itr.remove();
-
-                if (!this.processedDisconnect) {
-                    LOGGER.info("Disconnecting {} for sending keepalive response ({}) out-of-order!", this.playerProfile().name(), packet.getId());
-                    this.disconnectAsync(TIMEOUT_DISCONNECTION_MESSAGE, io.papermc.paper.connection.DisconnectionReason.TIMEOUT);
-                    return;
-                }
-                break;
-            }
-        }
-
-        if (!this.processedDisconnect) {
-            LOGGER.info("Disconnecting {} for sending keepalive response ({}) without matching challenge!", this.playerProfile().name(), packet.getId());
-            this.disconnectAsync(TIMEOUT_DISCONNECTION_MESSAGE, io.papermc.paper.connection.DisconnectionReason.TIMEOUT);
-            return;
+        if (this.keepAlivePending && packet.getId() == this.keepAliveChallenge) {
+            int i = (int)(Util.getMillis() - this.keepAliveTime);
+            this.latency = (this.latency * 3 + i) / 4;
+            this.keepAlivePending = false;
+        } else if (!this.isSingleplayerOwner()) {
+            this.disconnectAsync(TIMEOUT_DISCONNECTION_MESSAGE, io.papermc.paper.connection.DisconnectionReason.TIMEOUT); // Paper - add proper async disconnect
         }
-        // Paper end - improve keepalives
     }
 
     @Override
@@ -212,7 +191,7 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
         } else {
             bridge.removeChannel(channel);
         }
-    // Paper end
+        // Paper end
     }
 
     @Override
@@ -263,23 +242,20 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
     protected void keepConnectionAlive() {
         Profiler.get().push("keepAlive");
         long millis = Util.getMillis();
-        // Paper start - improve keepalives
-        if (this.checkIfClosed(millis) && !this.processedDisconnect) {
-            long currTime = System.nanoTime();
-
-            if ((currTime - this.keepAlive.lastKeepAliveTx) >= java.util.concurrent.TimeUnit.SECONDS.toNanos(1L)) {
-                this.keepAlive.lastKeepAliveTx = currTime;
-
-                io.papermc.paper.util.KeepAlive.PendingKeepAlive pka = new io.papermc.paper.util.KeepAlive.PendingKeepAlive(currTime, millis);
-                this.keepAlive.pendingKeepAlives.add(pka);
-                this.send(new ClientboundKeepAlivePacket(pka.challengeId()));
-            }
-
-            io.papermc.paper.util.KeepAlive.PendingKeepAlive oldest = this.keepAlive.pendingKeepAlives.peek();
-            if (oldest != null && (currTime - oldest.txTimeNS()) > java.util.concurrent.TimeUnit.MILLISECONDS.toNanos(KEEPALIVE_LIMIT)) {
-                LOGGER.info("{} was kicked due to keepalive timeout!", this.playerProfile().name());
-                this.disconnect(TIMEOUT_DISCONNECTION_MESSAGE, io.papermc.paper.connection.DisconnectionReason.TIMEOUT); // Paper - kick event cause
-                // Paper end - improve keepalives
+        // Paper start - give clients a longer time to respond to pings as per pre 1.12.2 timings
+        // This should effectively place the keepalive handling back to "as it was" before 1.12.2
+        final long elapsedTime = millis - this.keepAliveTime;
+        if (!this.isSingleplayerOwner() && elapsedTime >= 15000L) { // use vanilla's 15000L between keep alive packets
+            if (this.keepAlivePending) {
+                if (!this.processedDisconnect && elapsedTime >= KEEPALIVE_LIMIT) { // check keepalive limit, don't fire if already disconnected
+                    this.disconnect(TIMEOUT_DISCONNECTION_MESSAGE, io.papermc.paper.connection.DisconnectionReason.TIMEOUT); // Paper - kick event cause
+                }
+                // Paper end - give clients a longer time to respond to pings as per pre 1.12.2 timings
+            } else if (this.checkIfClosed(millis)) {
+                this.keepAlivePending = true;
+                this.keepAliveTime = millis;
+                this.keepAliveChallenge = millis;
+                this.send(new ClientboundKeepAlivePacket(this.keepAliveChallenge));
             }
         }
 
@@ -312,27 +288,33 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
     }
 
     public void send(Packet<?> packet, @Nullable ChannelFutureListener sendListener) {
-        // CraftBukkit start
-        if (packet == null || this.processedDisconnect) { // Spigot
+        if (packet == null || this.processedDisconnect) return;
+
+        // Pulse start - Smart Batching V3 (Delegation)
+        if (this.pulseBuffer != null) {
+            // We just send packet to buffer
+            this.pulseBuffer.add(packet, sendListener);
+
+            // If packet close connection send it as "Instant"
+            if (packet.isTerminal()) {
+                this.pulseBuffer.flush("Instant");
+                this.close();
+            }
             return;
-        } else if (packet instanceof net.minecraft.network.protocol.game.ClientboundSetDefaultSpawnPositionPacket defaultSpawnPositionPacket && this instanceof ServerGamePacketListenerImpl serverGamePacketListener) {
-            serverGamePacketListener.player.compassTarget = org.bukkit.craftbukkit.util.CraftLocation.toBukkit(defaultSpawnPositionPacket.respawnData().pos(), serverGamePacketListener.getPlayer().level());
-        }
-        // CraftBukkit end
-        if (packet.isTerminal()) {
-            this.close();
         }
 
-        boolean flag = !this.suspendFlushingOnServerThread || !this.server.isSameThread();
+        // Fallback if buffer doesnt created
+        if (packet instanceof net.minecraft.network.protocol.game.ClientboundSetDefaultSpawnPositionPacket p && this instanceof ServerGamePacketListenerImpl s) {
+            s.player.compassTarget = org.bukkit.craftbukkit.util.CraftLocation.toBukkit(p.respawnData().pos(), s.getPlayer().level());
+        }
+        if (packet.isTerminal()) this.close();
 
         try {
-            this.connection.send(packet, sendListener, flag);
+            this.connection.send(packet, sendListener, true);
         } catch (Throwable var7) {
-            CrashReport crashReport = CrashReport.forThrowable(var7, "Sending packet");
-            CrashReportCategory crashReportCategory = crashReport.addCategory("Packet being sent");
-            crashReportCategory.setDetail("Packet class", () -> packet.getClass().getCanonicalName());
-            throw new ReportedException(crashReport);
+            throw new net.minecraft.ReportedException(net.minecraft.CrashReport.forThrowable(var7, "Sending packet"));
         }
+        // Pulse end
     }
 
     @Deprecated @io.papermc.paper.annotation.DoNotUse // Paper - kick event causes
@@ -380,9 +362,9 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
             net.minecraft.server.level.ServerPlayer player = serverGamePacketListener.player;
             org.bukkit.event.player.PlayerKickEvent.Cause cause = disconnectionDetails.disconnectionReason().orElseThrow().game().orElse(org.bukkit.event.player.PlayerKickEvent.Cause.UNKNOWN);
             org.bukkit.event.player.PlayerKickEvent event = new org.bukkit.event.player.PlayerKickEvent(
-                    player.getBukkitEntity(),
-                    io.papermc.paper.adventure.PaperAdventure.asAdventure(disconnectionDetails.reason()),
-                    rawLeaveMessage, cause
+                player.getBukkitEntity(),
+                io.papermc.paper.adventure.PaperAdventure.asAdventure(disconnectionDetails.reason()),
+                rawLeaveMessage, cause
 
             );
 
@@ -415,10 +397,10 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
 
     private void disconnect0(DisconnectionDetails disconnectionDetails) {
         this.connection
-                .send(
-                        new ClientboundDisconnectPacket(disconnectionDetails.reason()),
-                        PacketSendListener.thenRun(() -> this.connection.disconnect(disconnectionDetails))
-                );
+            .send(
+                new ClientboundDisconnectPacket(disconnectionDetails.reason()),
+                PacketSendListener.thenRun(() -> this.connection.disconnect(disconnectionDetails))
+            );
         this.onDisconnect(disconnectionDetails);
         this.connection.setReadOnly();
         // CraftBukkit - Don't wait
@@ -460,6 +442,6 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
     }
 
     protected CommonListenerCookie createCookie(ClientInformation clientInformation) {
-        return new CommonListenerCookie(this.playerProfile(), this.latency, clientInformation, this.transferred, this.playerBrand, this.pluginMessagerChannels, this.keepAlive); // Paper
+        return new CommonListenerCookie(this.playerProfile(), this.latency, clientInformation, this.transferred, this.playerBrand, this.pluginMessagerChannels, this.keepAlive); // Paper // Pulse - add keepAlive
     }
 }
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 5c4dacec63cf40676ffc3cd28364377ae5408e35..b0e28d0edaa18cc9e7e336ce3203de533ffa4296 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -245,6 +245,7 @@ import org.bukkit.event.player.PlayerTeleportEvent;
 import org.bukkit.event.player.PlayerToggleFlightEvent;
 import org.bukkit.event.player.PlayerToggleSprintEvent;
 // CraftBukkit end
+// PULSE_MODIFIED
 
 public class ServerGamePacketListenerImpl
     extends ServerCommonPacketListenerImpl
@@ -366,6 +367,9 @@ public class ServerGamePacketListenerImpl
                 this.disconnect(Component.translatable("multiplayer.disconnect.idling"), org.bukkit.event.player.PlayerKickEvent.Cause.IDLING); // Paper - kick event cause
             }
         }
+        // Pulse start
+        if (this.pulseBuffer != null) this.pulseBuffer.flush("Tick");
+        // Pulse end
     }
 
     private boolean tickPlayer() {
@@ -649,7 +653,7 @@ public class ServerGamePacketListenerImpl
                     } // else: no collision at all detected, why do we care?
                 }
                 if (teleportBack) {
-                // Paper end - optimise out extra getCubes
+                    // Paper end - optimise out extra getCubes
                     rootVehicle.absSnapTo(x, y, z, f, f1);
                     this.send(ClientboundMoveVehiclePacket.fromEntity(rootVehicle));
                     rootVehicle.removeLatestMovementRecording();
@@ -1483,7 +1487,7 @@ public class ServerGamePacketListenerImpl
                                     this.lastTick = (int) (System.currentTimeMillis() / 50);
 
                                     if (i > Math.max(this.allowedPlayerTicks, 5)) {
-                                    // CraftBukkit end
+                                        // CraftBukkit end
                                         LOGGER.debug(
                                             "{} is sending move packets too frequently ({} packets since last tick)", this.player.getPlainTextName(), i
                                         );
@@ -1505,7 +1509,7 @@ public class ServerGamePacketListenerImpl
                                     if (this.player.level().paperConfig().chunks.preventMovingIntoUnloadedChunks && (this.player.getX() != toX || this.player.getZ() != toZ) && !serverLevel.areChunksLoadedForMove(this.player.getBoundingBox().expandTowards(new Vec3(toX, toY, toZ).subtract(this.player.position())))) {
                                         // Paper start - Add fail move event
                                         io.papermc.paper.event.player.PlayerFailMoveEvent event = fireFailMove(io.papermc.paper.event.player.PlayerFailMoveEvent.FailReason.MOVED_INTO_UNLOADED_CHUNK,
-                                                toX, toY, toZ, toYaw, toPitch, false);
+                                            toX, toY, toZ, toYaw, toPitch, false);
                                         if (!event.isAllowed()) {
                                             this.internalTeleport(PositionMoveRotation.of(this.player), Collections.emptySet());
                                             return;
@@ -1520,13 +1524,13 @@ public class ServerGamePacketListenerImpl
                                             // CraftBukkit end
                                             // Paper start - Add fail move event
                                             io.papermc.paper.event.player.PlayerFailMoveEvent event = fireFailMove(io.papermc.paper.event.player.PlayerFailMoveEvent.FailReason.MOVED_TOO_QUICKLY,
-                                                    toX, toY, toZ, toYaw, toPitch, true);
+                                                toX, toY, toZ, toYaw, toPitch, true);
                                             if (!event.isAllowed()) {
                                                 if (event.getLogWarning()) {
-                                            LOGGER.warn("{} moved too quickly! {},{},{}", this.player.getPlainTextName(), d3, d4, d5);
+                                                    LOGGER.warn("{} moved too quickly! {},{},{}", this.player.getPlainTextName(), d3, d4, d5);
                                                 }
                                                 this.teleport(
-                                                        this.player.getX(), this.player.getY(), this.player.getZ(), this.player.getYRot(), this.player.getXRot()
+                                                    this.player.getX(), this.player.getY(), this.player.getZ(), this.player.getYRot(), this.player.getXRot()
                                                 );
                                                 return;
                                             }
@@ -1598,12 +1602,12 @@ public class ServerGamePacketListenerImpl
                                     && !this.player.isInPostImpulseGraceTime()) {
                                     // Paper start - Add fail move event
                                     io.papermc.paper.event.player.PlayerFailMoveEvent event = fireFailMove(io.papermc.paper.event.player.PlayerFailMoveEvent.FailReason.MOVED_WRONGLY,
-                                            toX, toY, toZ, toYaw, toPitch, true);
+                                        toX, toY, toZ, toYaw, toPitch, true);
                                     if (!event.isAllowed()) {
                                         movedWrongly = true;
                                         if (event.getLogWarning())
-                                     // Paper end
-                                    LOGGER.warn("{} moved wrongly!", this.player.getPlainTextName());
+                                            // Paper end
+                                            LOGGER.warn("{} moved wrongly!", this.player.getPlainTextName());
                                     } // Paper
                                 }
 
@@ -1621,7 +1625,7 @@ public class ServerGamePacketListenerImpl
                                 // Paper end - optimise out extra getCubes
                                 if (!allowMovement) {
                                     io.papermc.paper.event.player.PlayerFailMoveEvent event = fireFailMove(io.papermc.paper.event.player.PlayerFailMoveEvent.FailReason.CLIPPED_INTO_BLOCK,
-                                            toX, toY, toZ, toYaw, toPitch, false);
+                                        toX, toY, toZ, toYaw, toPitch, false);
                                     if (event.isAllowed()) {
                                         allowMovement = true;
                                     }
@@ -2106,8 +2110,8 @@ public class ServerGamePacketListenerImpl
                                     && success.swingSource() == InteractionResult.SwingSource.SERVER && !this.player.gameMode.interactResult) { // Paper - Call interact event
                                     this.player.swing(hand, true);
                                 }
-                            // Paper start - Fix inventory desync; MC-99075
-                            // From serverLevel.mayInteract(this.player, blockPos)
+                                // Paper start - Fix inventory desync; MC-99075
+                                // From serverLevel.mayInteract(this.player, blockPos)
                             } else if (this.server.isUnderSpawnProtection(serverLevel, blockPos, player)) {
                                 if (io.papermc.paper.util.MCUtil.clientPredictsInteraction(this.player, serverLevel.getBlockState(blockPos), itemInHand)) {
                                     this.player.containerMenu.sendAllDataToRemote();
@@ -2738,32 +2742,32 @@ public class ServerGamePacketListenerImpl
         // CraftBukkit end
         // Paper start - Ensure that client receives chat packets in the same order that we add into the message signature cache
         synchronized (this.messageSignatureCache) {
-        this.send(
-            new ClientboundPlayerChatPacket(
-                this.nextChatIndex++,
-                chatMessage.link().sender(),
-                chatMessage.link().index(),
-                chatMessage.signature(),
-                chatMessage.signedBody().pack(this.messageSignatureCache),
-                chatMessage.unsignedContent(),
-                chatMessage.filterMask(),
-                boundChatType
-            )
-        );
-        MessageSignature messageSignature = chatMessage.signature();
-        if (messageSignature != null) {
-            this.messageSignatureCache.push(chatMessage.signedBody(), chatMessage.signature());
-            int i;
-            synchronized (this.lastSeenMessages) {
-                this.lastSeenMessages.addPending(messageSignature);
-                i = this.lastSeenMessages.trackedMessagesCount();
-            }
+            this.send(
+                new ClientboundPlayerChatPacket(
+                    this.nextChatIndex++,
+                    chatMessage.link().sender(),
+                    chatMessage.link().index(),
+                    chatMessage.signature(),
+                    chatMessage.signedBody().pack(this.messageSignatureCache),
+                    chatMessage.unsignedContent(),
+                    chatMessage.filterMask(),
+                    boundChatType
+                )
+            );
+            MessageSignature messageSignature = chatMessage.signature();
+            if (messageSignature != null) {
+                this.messageSignatureCache.push(chatMessage.signedBody(), chatMessage.signature());
+                int i;
+                synchronized (this.lastSeenMessages) {
+                    this.lastSeenMessages.addPending(messageSignature);
+                    i = this.lastSeenMessages.trackedMessagesCount();
+                }
 
-            if (i > 4096) {
-                this.disconnectAsync(Component.translatable("multiplayer.disconnect.too_many_pending_chats"), org.bukkit.event.player.PlayerKickEvent.Cause.TOO_MANY_PENDING_CHATS); // Paper - kick event cause & add proper async disconnect
+                if (i > 4096) {
+                    this.disconnectAsync(Component.translatable("multiplayer.disconnect.too_many_pending_chats"), org.bukkit.event.player.PlayerKickEvent.Cause.TOO_MANY_PENDING_CHATS); // Paper - kick event cause & add proper async disconnect
+                }
             }
         }
-        }
         // Paper end - Ensure that client receives chat packets in the same order that we add into the message signature cache
     }
 
@@ -3654,12 +3658,12 @@ public class ServerGamePacketListenerImpl
             .append(
                 () -> {
                     server.executeBlocking(() -> { // Paper - Broadcast chat session update sync
-                    this.player.setChatSession(chatSession);
-                    this.server
-                        .getPlayerList()
-                        .broadcastAll(
-                            new ClientboundPlayerInfoUpdatePacket(EnumSet.of(ClientboundPlayerInfoUpdatePacket.Action.INITIALIZE_CHAT), List.of(this.player)), this.player // Paper - Use single player info update packet on join
-                        );
+                        this.player.setChatSession(chatSession);
+                        this.server
+                            .getPlayerList()
+                            .broadcastAll(
+                                new ClientboundPlayerInfoUpdatePacket(EnumSet.of(ClientboundPlayerInfoUpdatePacket.Action.INITIALIZE_CHAT), List.of(this.player)), this.player // Paper - Use single player info update packet on join
+                            );
                     });
                 }
             );
