From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ExTr1se <160177189+N1f-Angel@users.noreply.github.com>
Date: Thu, 5 Feb 2026 13:01:21 +0200
Subject: [PATCH] Pulse Optimizations: PooledByteBuf


diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index 4ed9611994c5c8da01fede690197527c5b3a5731..f38564699af4dddd1172786ce886d16b77e34b44 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -5,6 +5,7 @@ import com.google.common.collect.Queues;
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import com.mojang.logging.LogUtils;
 import io.netty.bootstrap.Bootstrap;
+import io.netty.buffer.PooledByteBufAllocator;
 import io.netty.channel.Channel;
 import io.netty.channel.ChannelException;
 import io.netty.channel.ChannelFuture;
@@ -698,7 +705,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
             eventLoopGroup = NETWORK_WORKER_GROUP.get();
         }
 
-        return new Bootstrap().group(eventLoopGroup).handler(new ChannelInitializer<Channel>() {
+        return new Bootstrap().option(io.netty.channel.ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT).group(eventLoopGroup).handler(new ChannelInitializer<Channel>() {
             @Override
             protected void initChannel(Channel channel) {
                 try {
@@ -761,7 +768,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 
     public static Connection connectToLocalServer(SocketAddress address) {
         final Connection connection = new Connection(PacketFlow.CLIENTBOUND);
-        new Bootstrap().group(LOCAL_WORKER_GROUP.get()).handler(new ChannelInitializer<Channel>() {
+        new Bootstrap().option(io.netty.channel.ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT).group(LOCAL_WORKER_GROUP.get()).handler(new ChannelInitializer<Channel>() {
             @Override
             protected void initChannel(Channel channel) {
                 ChannelPipeline channelPipeline = channel.pipeline();
diff --git a/net/minecraft/server/network/ServerConnectionListener.java b/net/minecraft/server/network/ServerConnectionListener.java
index bd07e6a5aa1883786d789ea71711a0c0c0a95c26..ceac63704931f996a9806f529c188458db942100 100644
--- a/net/minecraft/server/network/ServerConnectionListener.java
+++ b/net/minecraft/server/network/ServerConnectionListener.java
@@ -5,6 +5,7 @@ import com.google.common.collect.Lists;
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import com.mojang.logging.LogUtils;
 import io.netty.bootstrap.ServerBootstrap;
+import io.netty.buffer.PooledByteBufAllocator;
 import io.netty.channel.Channel;
 import io.netty.channel.ChannelException;
 import io.netty.channel.ChannelFuture;
@@ -117,6 +128,7 @@ public class ServerConnectionListener {
                 .add(
                     new ServerBootstrap()
                         .channel(clazz)
+                        .childOption(io.netty.channel.ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT)
                         .childHandler(
                             new ChannelInitializer<Channel>() {
                                 @Override
@@ -200,7 +212,8 @@ public class ServerConnectionListener {
         synchronized (this.channels) {
             channelFuture = new ServerBootstrap()
                 .channel(LocalServerChannel.class)
-                .childHandler(
+                .childOption(io.netty.channel.ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT)
+                .childHandler(
                     new ChannelInitializer<Channel>() {
                         @Override
                         protected void initChannel(Channel channel) {
@@ -211,6 +224,10 @@ public class ServerConnectionListener {
                             channelPipeline.addLast("packet_handler", connection);
                         }
                     });
+            this.channels.add(channelFuture);
+        }
+        return channelFuture.address();
+    }
--
