From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: tizis0 <hexgames84@gmail.com>
Date: Thu, 15 Jan 2026 23:24:39 +0600
Subject: [PATCH] Add '/pulse bar' for real-time metrics


diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index c8e36a9f1e5043b0e74a28d5886882d98b641631..30d374d31e986fd1996f61a1b8454e0be5cf4870 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -319,6 +319,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         // Pulse start - Init Config and Commands
         dev.pulsemc.config.ConfigManager.load();
         dev.pulsemc.network.PulseMetrics.start();
+        dev.pulsemc.network.PulseBar.start();
         this.server.getCommandMap().register("pulse", "Pulse", new org.bukkit.command.defaults.BukkitCommand("pulse") {
             private final dev.pulsemc.command.PulseCommand executor = new dev.pulsemc.command.PulseCommand();
             @Override
