--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -151,6 +_,7 @@
 import org.jetbrains.annotations.Contract;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
+// PULSE_MODIFIED
 
 public abstract class Entity implements SyncedDataHolder, DebugValueSource, Nameable, ItemOwner, SlotProvider, EntityAccess, ScoreHolder, DataComponentGetter, ca.spottedleaf.moonrise.patches.chunk_system.entity.ChunkSystemEntity, ca.spottedleaf.moonrise.patches.entity_tracker.EntityTrackerEntity { // Paper - rewrite chunk system // Paper - optimise entity tracker
     public static javax.script.ScriptEngine scriptEngine = new javax.script.ScriptEngineManager().getEngineByName("rhino"); // Purpur - Configurable entity base attributes
@@ -320,6 +_,13 @@
     private boolean invulnerable;
     protected UUID uuid; // Purpur - Add toggle for RNG manipulation
     protected String stringUUID; // Purpur - Add toggle for RNG manipulation
+    // Pulse start - connect Virtual Entities
+    @javax.annotation.Nullable
+    public final java.util.Set<java.util.UUID> pulseVirtualViewers = new java.util.concurrent.CopyOnWriteArraySet<>();
+    public boolean pulseIsVirtual() {
+        return this.pulseVirtualViewers != null && !this.pulseVirtualViewers.isEmpty();
+    }
+    // Pulse end - connect Virtual Entities
     private boolean hasGlowingTag;
     private final Set<String> tags = new io.papermc.paper.util.SizeLimitedSet<>(new it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<>(), MAX_ENTITY_TAG_COUNT); // Paper - fully limit tag size - replace set impl
     private final double[] pistonDeltas = new double[]{0.0, 0.0, 0.0};
@@ -1912,6 +_,20 @@
     }
 
     public void playSound(SoundEvent sound, float volume, float pitch) {
+        // Pulse start - connect Virtual Entities
+        if (this.pulseIsVirtual()) {
+            org.bukkit.Sound bukkitSound = org.bukkit.craftbukkit.CraftSound.minecraftToBukkit(sound);
+            org.bukkit.SoundCategory cat = org.bukkit.SoundCategory.valueOf(this.getSoundSource().name());
+            assert this.pulseVirtualViewers != null;
+            for (java.util.UUID uuid : this.pulseVirtualViewers) {
+                org.bukkit.entity.Player p = org.bukkit.Bukkit.getPlayer(uuid);
+                if (p != null) {
+                    p.playSound(this.getBukkitEntity().getLocation(), bukkitSound, cat, volume, pitch);
+                }
+            }
+            return;
+        }
+        // Pulse end - connect Virtual Entities
         if (!this.isSilent()) {
             this.level().playSound(null, this.getX(), this.getY(), this.getZ(), sound, this.getSoundSource(), volume, pitch);
         }
@@ -2323,6 +_,27 @@
     }
 
     public void push(Entity entity) {
+        // Pulse start - connect Virtual Entities
+        if (this.pulseIsVirtual() || entity.pulseIsVirtual()) {
+            if (this.pulseIsVirtual() && entity.pulseIsVirtual()) {
+                boolean shared = false;
+                for (java.util.UUID u : this.pulseVirtualViewers) {
+                    if (entity.pulseVirtualViewers.contains(u)) { shared = true; break; }
+                }
+                if (!shared) return;
+            }
+            else if (this.pulseIsVirtual() && entity instanceof net.minecraft.server.level.ServerPlayer p) {
+                if (!this.pulseVirtualViewers.contains(p.getUUID())) return;
+            }
+            else if (entity.pulseIsVirtual() && this instanceof net.minecraft.server.level.ServerPlayer p) {
+                if (!entity.pulseVirtualViewers.contains(p.getUUID())) return;
+            }
+            else {
+                return;
+            }
+        }
+        // Pulse end - connect Virtual Entities
+
         if (!this.isPassengerOfSameVehicle(entity)) {
             if (!entity.noPhysics && !this.noPhysics) {
                 if (this.level.paperConfig().collisions.onlyPlayersCollide && !(entity instanceof ServerPlayer || this instanceof ServerPlayer)) return; // Paper - Collision option for requiring a player participant
@@ -2549,6 +_,7 @@
     }
 
     public boolean save(ValueOutput output) {
+        if (this.pulseIsVirtual()) return false; // Pulse - Prevent virtual entities from saving
         return !this.isPassenger() && this.saveAsPassenger(output);
     }
 
@@ -4312,7 +_,10 @@
     }
 
     public boolean isIgnoringBlockTriggers() {
-        return false;
+        // Pulse start - connect Virtual Entities
+        return this.pulseIsVirtual();
+//        return false;
+        // Pulse end - connect Virtual Entities
     }
 
     public void fillCrashReportCategory(CrashReportCategory category) {
